Add initialization and error handling (#3632)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/error_handling.rst

.. meta::
   :description: Error Handling
   :keywords: AMD, ROCm, HIP, error handling, error

********************************************************************************
Error handling
********************************************************************************

HIP provides functionality to detect, report, and manage errors that occur
during the execution of HIP runtime functions or when launching kernels. Every
HIP runtime function, apart from launching kernels, has :cpp:type:`hipError_t`
as return type. :cpp:func:`hipGetLastError()` and :cpp:func:`hipPeekAtLastError()`
can be used for catching errors from kernel launches, as kernel launches don't
return an error directly. HIP maintains an internal state, that includes the
last error code. hipGetLastError returns and resets that error to hipSuccess,
while :cpp:func:`hipPeekAtLastError` just returns the error without changing it.
To get a human readable version of the errors, :cpp:func:`hipGetErrorString()`
and :cpp:func:`hipGetErrorName()` can be used.

> 

Reviewer: Currently, there is a slight difference in the behavior of hipGetLastError when compared with cudaGetLastError.  
- hipGetLastError returns the returned error code of just the last API even if its hipSuccess.
- cudaGetLastError returns the error returned by any of the preceeding CUDA APIs in the same host thread. 

HIP behavior will be matched with CUDA in the next major ROCm release 7.0.  Meanwhile we are planning to introduce an environment variable in 6.4, which when enabled will match the CUDA behavior, otherwise it will continue with the current behavior by default.

Please check if we can mention this difference of the API behavior in this document.

Reviewer 2: Added note.

.. meta::
   :description: Error Handling
   :keywords: AMD, ROCm, HIP, error handling, error

********************************************************************************
Error handling
********************************************************************************

> HIP provides functionality to detect, report, and manage errors that occur

Reviewer 3: I think I would combine this first paragraph with the bulleted list of Important Error Handling functions below. Maybe start with the Error Handling Usage as the Intro?

Reviewer 2: I combined the paragraphs.

during the execution of HIP runtime functions or when launching kernels. Every
HIP runtime function, apart from launching kernels, has :cpp:type:`hipError_t`
as return type. :cpp:func:`hipGetLastError()` and :cpp:func:`hipPeekAtLastError()`
can be used for catching errors from kernel launches, as kernel launches don't
return an error directly. HIP maintains an internal state, that includes the
last error code. hipGetLastError returns and resets that error to hipSuccess,
while :cpp:func:`hipPeekAtLastError` just returns the error without changing it.
To get a human readable version of the errors, :cpp:func:`hipGetErrorString()`
and :cpp:func:`hipGetErrorName()` can be used.

Error handling usage
================================================================================

> Error handling functions enable developers to implement appropriate error

Reviewer 3: ```suggestion
Error handling functions enable developers to implement appropriate strategies
```

> handling strategies, such as retry mechanisms, resource cleanup, or graceful

Reviewer 3: ```suggestion
for handling errors, such as retry mechanisms, resource cleanup, or graceful
```

degradation.

> The descriptions of the important :ref:`error handling functions <error_handling_reference>`

Reviewer 3: ```suggestion
Descriptions of the important :ref:`error handling functions <error_handling_reference>`
```

> are the following:

Reviewer 3: ```suggestion
are as follows:
```

.. meta::
   :description: Error Handling
   :keywords: AMD, ROCm, HIP, error handling, error

********************************************************************************
Error handling
********************************************************************************

The Error handling API in HIP provides the necessary tools to detect, report,

> and manage errors in GPU-accelerated applications. By checking return values,

Reviewer 4: This should emphasize, that it is only related to HIP runtime errors (and kernels). It should also explicitly mention, that HIP runtime functions return an error code.
```suggestion
HIP provides functionality to detect, report, and manage errors that occur during the execution of HIP runtime functions or when launching kernels. Every HIP runtime function, apart from launching kernels, has `hipError_t` as return type.
```

using functions like :cpp:func:`hipGetErrorString()`, :cpp:func:`hipGetLastError()`,

> and :cpp:func:`hipPeekAtLastError()`, and adopting best practices like defining

Reviewer 4: This is mixing two different aspects of error handling - GetErrorString/Name are for making it human readable, the Get/PeekAtLastError are for "gathering" errors. I know the functionality of Get/PeekAtLastError is explained later, but I think it's fundamental enough to explain it in the introduction.
```suggestion
:cpp:func:`hipGetLastError()` and :cpp:func:`hipPeekAtLastError()` can be used for catching errors from kernel launches, as kernel launches don't return an error directly. HIP maintains an internal state, that includes the last error code. hipGetLastError returns and resets that error to hipSuccess, while hipPeekAtLastError just returns the error without changing it.
To get a human readable version of the errors, :cpp:func:`hipGetErrorString()` and hipGetErrorName() can be used.
```

error-checking macros, developers can ensure their HIP applications are robust,

> easier to debug, and more reliable. Proper error handling is crucial for

Reviewer 4: Error handling doesn't make an application more robust or reliable by itself. There are some errors that can't be recovered from like running out of memory for example.

Reviewer 2: Fixed.

identifying issues early in the development process and ensuring that
applications behave as expected.

Strategies

> ================================================================================

Reviewer 4: This section feels a bit out of place - it is not HIP specific, and is fairly vague in general. I would vote for removing it, as it doesn't add much to the topic at hand, or at least move it to the end of the page as an "addendum"

Reviewer 2: Removed strategies.

One of the fundamental best practices in error handling is to develop a
consistent strategy across the entire application. This involves defining
how errors are reported, logged, and managed. This can be achieved by using a
centralized error handling mechanism that ensures consistency and reduces

> redundancy. For instance, using macros to simplify error checking and reduce

Reviewer 4: This is a whole lot of buzzwords without saying much - ensuring what kind of consistency? reducing what redundancy?

Reviewer 2: Removed strategies.

code duplication is a common practice. A macro like ``HIP_CHECK`` can be defined

> to check the return value of HIP API calls and handle errors appropriately.

Reviewer 4: HIP_CHECK is actually not used for handling errors at all, it is just reporting errors, and in most implementations of it I've seen just terminates function or even program execution.

Also referring to HIP_CHECK without giving some example code might confuse users - HIP_CHECK is not a macro that is defined by the HIP runtime, it is a custom macro that the users have to define themselves.

Reviewer 2: Removed strategies.

Granular error reporting
--------------------------------------------------------------------------------

It involves reporting errors at the appropriate level of detail. Too much detail
can overwhelm users, while too little can make debugging difficult.
Differentiating between user-facing errors and internal errors is crucial. 

Fail-fast principle
--------------------------------------------------------------------------------

It involves detecting and handling errors as early as possible to prevent them
from propagating and causing more significant issues. Such as validating inputs
and preconditions before performing operations.

Resource management 
--------------------------------------------------------------------------------

Ensuring that resources such as memory, file handles, and network connections

> are properly managed and released in the event of an error is essential.

Reviewer 4: This is not really a error handling strategy, it's not like this is optional. If this is not done, it can e.g. corrupt data

Reviewer 2: Removed strategies.

Error handling usage
================================================================================

> Functions like :cpp:func:`hipGetLastError` and :cpp:func:`hipPeekAtLastError`

Reviewer 4: Those are the only two functions used for that purpose, so no "like" is needed. (When ignoring the driver API versions of those functions)
```suggestion
:cpp:func:`hipGetLastError` and :cpp:func:`hipPeekAtLastError`
```

are used to detect errors after HIP API calls. This ensures that any issues are

> caught early in the execution flow.

Reviewer 4: These functions do not ensure that errors are caught early - actually the opposite. They should be used for kernel launches, which don't return an error code, or asynchronous functions, which only return a subset of possible errors when called. The device-related errors of the asynchronous functions can only be caught later, as they are... well, asynchronous.

Reviewer 2: Rewrote.

For reporting, :cpp:func:`hipGetErrorName` and :cpp:func:`hipGetErrorString`

> provide meaningful error messages that can be logged or displayed to users.

Reviewer 4: This depends on what you understand by "logging". An internal logging mechanism doesn't need the human readable presentation of the errors, in that case it would be easier to use the error codes.

Reviewer 2: Rewrote.

This helps in understanding the nature of the error and facilitates debugging.

By checking for errors and providing detailed information, these functions
enable developers to implement appropriate error handling strategies, such as
retry mechanisms, resource cleanup, or graceful degradation.

Error handling function examples

> --------------------------------------------------------------------------------

Reviewer 4: I think these functions are all simple enough, that they don't need code examples - it basically boils down to calling the function. And e.g. hipGetErrorString has already been used twice in the code examples before its example code is introduced.

In my opinion it would be better to replace this section with the infamous HIP_CHECK macro and highlight that and its usage a bit more. After all, that is almost the first thing useres (should) learn when starting to code with HIP, it might as well be part of the "hello world" version of HIP

:cpp:func:`hipGetLastError` returns the last error that occurred during a HIP
runtime API call and resets the error code to :cpp:enumerator:`hipSuccess`:

.. code-block:: cpp

  hipError_t err = hipGetLastError();
  if (err != hipSuccess)
  {
      std::cerr << "HIP error: " << hipGetErrorString(err) << std::endl;
  }

:cpp:func:`hipPeekAtLastError` returns the last error that occurred during a HIP
runtime API call **without** resetting the error code:

.. code-block:: cpp

  hipError_t err = hipPeekAtLastError();
  if (err != hipSuccess)
  {
      std::cerr << "HIP error: " << hipGetErrorString(err) << std::endl;
  }

:cpp:func:`hipGetErrorName` converts a HIP error code to a string representing
the error name:

.. code-block:: cpp

  std::cerr << "Error name: " << hipGetErrorName(err) << std::endl;

:cpp:func:`hipGetErrorString` converts a HIP error code to a string describing
the error:

.. code-block:: cpp

  std::cerr << "Error description: " << hipGetErrorString(err) << std::endl;

Best practices
--------------------------------------------------------------------------------

1. Check errors after each API call

>   Always check the return value of HIP API calls to catch errors early. For

Reviewer 4: The emphasis on catching errors "early" is a bit out of place - it's as if it is an option to not catch errors early, and continue execution as if nothing has happened

  example:

  .. code-block:: cpp

    hipError_t err = hipMalloc(&d_A, size);
    if (err != hipSuccess) {
        std::cerr << "hipMalloc failed: " << hipGetErrorString(err) << std::endl;
        return -1;
    }

> 2. Use macros for error checking

Reviewer 4: As already mentioned: this should get its own section. It is a very important concept for begginners. However it should be mentioned, that this is only reporting errors, not handling them. There actually are versions of this macro, that have a flag for exiting the program after reporting the error.

        }                                          \
    }

    // Allocate memory on the device with error check.
    HIP_CHECK(hipMalloc(&device_pointer, N * sizeof(float);));

3. Handle errors gracefully

   Ensure the application can handle errors gracefully, such as by freeing
   resources or providing meaningful error messages to the user.

Complete example
--------------------------------------------------------------------------------

A complete example demonstrating error handling:

.. code-block:: cpp

  #include <hip/hip_runtime.h>
  #include <iostream>

  #define HIP_CHECK(expression)                  \
  {                                              \
      const hipError_t status = expression;      \
      if(status != hipSuccess){                  \
          std::cerr << "HIP error "              \
                    << status << ": "            \
                    << hipGetErrorString(status) \
                    << " at " << __FILE__ << ":" \
                    << __LINE__ << std::endl;    \
      }                                          \
  }

  // Addition of two values.
  __global__ void add(int *a, int *b, int *c) {
      *c = *a + *b;

>   }

Reviewer 4: Even if the focus here is on the error handling, I think we shouldn't encourage people to program kernels this way. With some slight changes, this could be a vector(array)-add-like kernel. Then we also wouldn't allocate memory for "scalar" variables, and wouldn't launch a kernel with 1 thread in 1 block

Reviewer 2: I disagree with this. I do not think the 1 thread in 1 block kernel sample code is encouraging people to a bad habit.

Obviously no one should add two values at kernel side, but for showing the error handling where the kernel is not the main focus of the example code can help to understand the error handling faster.

Reviewer 4: I didn't mean the performance of this kernel, or whether it even is suited for GPUs, I wouldn't even mind if it was a only the initialization of an array, even those kernels can have their justifications on GPUs, if they in turn avoid a copy from host to device.

This code doesn't provide any parallelism, even if it was launched with more threads, as it does not use threadIds etc. Expanding this to properly use several threads wouldn't expand the code by too much, it mostly would be adding a "array_size" variable, that would be added to the malloc and memcpy calls, and would add two lines overall to the kernel - getting the threads index and checking the bounds of the array. I don't think that would distract the user from the error handling in this example.

Reviewer 2: You got your size.

  int main() {
      int a, b, c;
      int *d_a, *d_b, *d_c;

      // Setup input values.
      a = 1;
      b = 2;

      // Allocate device copies of a, b and c.
      HIP_CHECK(hipMalloc(&d_a, sizeof(*d_a)));
      HIP_CHECK(hipMalloc(&d_b, sizeof(*d_b)));
      HIP_CHECK(hipMalloc(&d_c, sizeof(*d_c)));

      // Copy input values to device.
      HIP_CHECK(hipMemcpy(d_a, &a, sizeof(*d_a), hipMemcpyHostToDevice));
      HIP_CHECK(hipMemcpy(d_b, &b, sizeof(*d_b), hipMemcpyHostToDevice));

      // Launch add() kernel on GPU.
      hipLaunchKernelGGL(add, dim3(1), dim3(1), 0, 0, d_a, d_b, d_c);
      // Check the kernel launch

>       HIP_CHECK(hipPeekAtLastError());

Reviewer 4: Unless there is a reason for not resetting the error, I think hipGetLastError should be used. This way, if there was some kind of error handling, the subsequent function might fail with the same error again.
```suggestion
      HIP_CHECK(hipGetLastError());
```

      // Check for kernel execution error
      HIP_CHECK(hipDeviceSynchronize());
      

>       //Check the kernel launch
>       HIP_CHECK(hipPeekAtLastError());

Reviewer 4: The kernel launch has already been checked
```suggestion
```

      // Copy the result back to the host.
      HIP_CHECK(hipMemcpy(&c, d_c, sizeof(*d_c), hipMemcpyDeviceToHost));

      // Cleanup allocated memory.
      HIP_CHECK(hipFree(d_a));
      HIP_CHECK(hipFree(d_b));
      HIP_CHECK(hipFree(d_c));

>       // Prints the result.

Reviewer 4: Being pedantic, sorry :) same style as the other comments:
```suggestion
      // Print the result.
```
